contrib/pkg/adm/managedns/enable.go

```diff
@@ -160,112 +171,159 @@ func (o *Options) Run(dynClient client.Client, args []string) error {
+
+	unstDeployment, err := dynClient.Resource(gvr).Namespace(constants.HiveNamespace).Get(hiveAdmissionDeployment, metav1.GetOptions{})
+	if err != nil {
+		log.WithError(err).Error("failed to get deployment object")
+		return err
+	}
+
+	deployment := &appsv1.Deployment{}
+	if err := scheme.Scheme.Convert(unstDeployment, deployment, nil); err != nil {
+		log.WithError(err).Error("failed to convert unstructed deployment to deployment")
+		return err
+	}
+
+	statusViewer := &kubectl.DeploymentStatusViewer{}
+
+	fieldSelector := fields.OneTermEqualSelector("metadata.name", hiveAdmissionDeployment).String()
+
+	lw := &cache.ListWatch{
+		ListFunc: func(options metav1.ListOptions) (runtime.Object, error) {
+			options.FieldSelector = fieldSelector
+			return dynClient.Resource(gvr).Namespace(constants.HiveNamespace).List(options)
+
+		},
+		WatchFunc: func(options metav1.ListOptions) (watch.Interface, error) {
+			options.FieldSelector = fieldSelector
+			return dynClient.Resource(gvr).Namespace(constants.HiveNamespace).Watch(options)
+		},
+	}
+
+	preconditionFunc := func(store cache.Store) (bool, error) {
+		_, exists, err := store.Get(&metav1.ObjectMeta{Namespace: constants.HiveNamespace, Name: hiveAdmissionDeployment})
 		if err != nil {
-			log.WithError(err).Fatal("error listing hiveadmission pods")
-		}
-		podsReady := true
-		for _, p := range hiveAdmissionPods.Items {
-			if !podUsingConfigMap(p, dnsConfigMapName) {
-				podsReady = false
-				break
-			}
+			return true, err
 		}
-		if !podsReady {
-			log.Info("new hiveadmission pods not ready, sleeping 5s")
-			time.Sleep(5 * time.Second)
+		if !exists {
+			// We need to make sure we see the object in the cache before we start waiting for events
+			// or we would be waiting for the timeout if such object didn't exist.
+			return true, apierrors.NewNotFound(appsv1.Resource(resourceName), hiveAdmissionDeployment)
 		}
+
+		return false, nil
 	}
 
-	return nil
-}
+	ctx, cancel := context.WithTimeout(context.Background(), waitTime)
+	defer cancel()
 
-func waitForHiveConfigToBeProcessed(dynClient client.Client) error {
-	// Spin while we wait for the HiveConfig.Status.LastObservedGeneration to match
-	// the HiveConfig's Generation.
-
-	key := types.NamespacedName{Namespace: constants.HiveNamespace, Name: hiveConfigName}
-
-	log.Info("waiting for new HiveConfig to be processed")
-	for i := 0; i < 24; i++ {
-		hc := &hivev1.HiveConfig{}
-		err := dynClient.Get(context.TODO(), key, hc)
-		switch {
-		case err != nil:
-			log.WithError(err).Error("error getting HiveConfig object")
-			return err
-		case hc.Generation == hc.Status.LastGenerationObserved && hc.Status.ConfigApplied:
-			return nil
-		}
+	_, err = clientwatch.UntilWithSync(ctx, lw, &unstructured.Unstructured{}, preconditionFunc, func(e watch.Event) (bool, error) {
+		switch t := e.Type; t {
+		case watch.Added, watch.Modified:
+			_, done, err := statusViewer.Status(e.Object.(runtime.Unstructured), deployment.Generation)
```

Comment:
I would pass in a revision of 0 instead of the revision from the deployment. We do not care about reaching that particular revision: We only care about reaching a point where the observedGeneration and generation are equal.

This also means that we do not need to get the deployment prior to the call to `UntilWithSync`.